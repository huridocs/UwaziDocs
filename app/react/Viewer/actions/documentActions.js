import api from 'app/utils/api';
import referencesAPI from 'app/Viewer/referencesAPI';
import * as types from 'app/Viewer/actions/actionTypes';
import * as connectionsTypes from 'app/Connections/actions/actionTypes';
import { entityDefaultDocument } from 'shared/entityDefaultDocument';

import { APIURL } from 'app/config.js';
import { actions } from 'app/BasicReducer';
import { actions as formActions } from 'react-redux-form';
import { documentsApi } from 'app/Documents';
import { notificationActions } from 'app/Notifications';
import { removeDocument, unselectAllDocuments } from 'app/Library/actions/libraryActions';
import { isClient } from 'app/utils';
import { actions as relationshipActions } from 'app/Relationships';
import { RequestParams } from 'app/utils/RequestParams';
import * as selectionActions from './selectionActions';
import * as uiActions from './uiActions';
import { PDFUtils } from '../../PDF';

export function setDocument(document, html) {
  return {
    type: types.SET_DOCUMENT,
    document,
    html,
  };
}

export function resetDocumentViewer() {
  return {
    type: types.RESET_DOCUMENT_VIEWER,
  };
}

export function loadDefaultViewerMenu() {
  return {
    type: types.LOAD_DEFAULT_VIEWER_MENU,
  };
}

export function saveDocument(doc) {
  const updateDoc = {};
  Object.keys(doc).forEach(key => {
    if (key !== 'fullText') {
      updateDoc[key] = doc[key];
    }
  });

  return dispatch =>
    documentsApi.save(new RequestParams(updateDoc)).then(updatedDoc => {
      dispatch(notificationActions.notify('Document updated', 'success'));
      dispatch({ type: types.VIEWER_UPDATE_DOCUMENT, doc });
      dispatch(formActions.reset('documentViewer.sidepanel.metadata'));
      dispatch(actions.update('viewer/doc', updatedDoc));
      dispatch(relationshipActions.reloadRelationships(updatedDoc.sharedId));
    });
}

export function saveToc(toc, fileId) {
  return async (dispatch, getState) => {
    const currentDoc = getState().documentViewer.doc.toJS();
    dispatch(formActions.reset('documentViewer.sidepanel.metadata'));
    dispatch(actions.set('documentViewer/tocBeingEdited', false));

    const updatedFile = (await api.post('files', new RequestParams({ toc, _id: fileId }))).json;
    const doc = {
      ...currentDoc,
      defaultDoc: updatedFile,
      documents: currentDoc.documents.map(d => {
        if (d._id === updatedFile._id) {
          return updatedFile;
        }
        return d;
      }),
    };

    dispatch(notificationActions.notify('Document updated', 'success'));
    dispatch({ type: types.VIEWER_UPDATE_DOCUMENT, doc });
    dispatch(formActions.reset('documentViewer.sidepanel.metadata'));
    dispatch(actions.set('viewer/doc', doc));
  };
}

export function deleteDocument(doc) {
  return dispatch =>
    documentsApi.delete(new RequestParams({ sharedId: doc.sharedId })).then(() => {
      dispatch(notificationActions.notify('Document deleted', 'success'));
      dispatch(resetDocumentViewer());
      dispatch(removeDocument(doc));
      dispatch(unselectAllDocuments());
    });
}

export async function getDocument(requestParams, defaultLanguage, filename) {
  const [entity] = (await api.get('entities', requestParams)).json.rows;

  const defaultDoc = filename
    ? entity.documents.find(d => d.filename === filename)
    : entityDefaultDocument(entity.documents, entity.language, defaultLanguage);

  entity.defaultDoc = defaultDoc;
  if (!isClient) {
    return entity;
  }
  if (defaultDoc.pdfInfo) {
    return entity;
  }

  const pdfInfo = await PDFUtils.extractPDFInfo(`${APIURL}files/${defaultDoc.filename}`);
  const processedDoc = await api
    .post('documents/pdfInfo', new RequestParams({ _id: defaultDoc._id, pdfInfo }))
    .then(res => res.json);

  return {
    ...entity,
    defaultDoc: processedDoc,
    documents: entity.documents.map(d => {
      if (d._id === processedDoc._id) {
        return processedDoc;
      }
      return d;
    }),
  };
}

export function loadTargetDocument(sharedId) {
  return (dispatch, getState) =>
    Promise.all([
      getDocument(new RequestParams({ sharedId }), getState().locale),
      referencesAPI.get(new RequestParams({ sharedId })),
    ]).then(([targetDoc, references]) => {
      dispatch(actions.set('viewer/targetDoc', targetDoc));
      dispatch(actions.set('viewer/targetDocReferences', references));
    });
}

export function cancelTargetDocument() {
  return dispatch => {
    dispatch({ type: connectionsTypes.CANCEL_RANGED_CONNECTION });
    dispatch(actions.unset('viewer/targetDoc'));
    dispatch(actions.unset('viewer/targetDocReferences'));
    dispatch(selectionActions.unsetTargetSelection());
    dispatch(uiActions.openPanel('viewMetadataPanel'));
  };
}

export function editToc(toc) {
  return dispatch => {
    dispatch(actions.set('documentViewer/tocBeingEdited', true));
    dispatch(formActions.load('documentViewer.tocForm', toc));
    dispatch(uiActions.openPanel('viewMetadataPanel'));
    dispatch(actions.set('viewer.sidepanel.tab', 'toc'));
  };
}

export function removeFromToc(tocElement) {
  return (dispatch, getState) => {
    const state = getState();
    let toc = state.documentViewer.tocForm;

    toc = toc.filter(entry => entry !== tocElement);

    dispatch(formActions.load('documentViewer.tocForm', toc));
  };
}

export function indentTocElement(tocElement, indentation) {
  return (dispatch, getState) => {
    const state = getState();
    const toc = state.documentViewer.tocForm.map(_entry => {
      const entry = Object.assign({}, _entry);
      if (_entry === tocElement) {
        entry.indentation = indentation;
      }
      return entry;
    });

    dispatch(formActions.load('documentViewer.tocForm', toc));
  };
}

export function addToToc(textSelectedObject, currentToc) {
  return (dispatch, getState) => {
    const state = getState();
    let toc = state.documentViewer.tocForm.concat();
    if (!toc.length) {
      toc = currentToc;
    }
    const tocElement = {
      range: {
        start: textSelectedObject.sourceRange.start,
        end: textSelectedObject.sourceRange.end,
      },
      label: textSelectedObject.sourceRange.text,
      indentation: 0,
    };

    toc.push(tocElement);
    toc = toc.sort((a, b) => a.range.start - b.range.start);
    dispatch(editToc(toc));
  };
}
