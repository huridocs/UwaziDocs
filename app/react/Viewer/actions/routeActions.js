import referencesAPI from 'app/Viewer/referencesAPI';
import relationTypesAPI from 'app/RelationTypes/RelationTypesAPI';
import referencesUtils from 'app/Viewer/utils/referencesUtils';
import {getDocument} from 'app/Viewer/actions/documentActions';
import {get as prioritySortingCriteria} from 'app/utils/prioritySortingCriteria';

import {actions} from 'app/BasicReducer';
import {setReferences} from './referencesActions';
import {actions as formActions} from 'react-redux-form';

export function requestViewerState(documentId, lang, globalResources) {
  return Promise.all([
    getDocument(documentId),
    referencesAPI.get(documentId),
    relationTypesAPI.get(),
    // TEST!!!
    referencesAPI.getGroupedByConnection(documentId)
  ])
  .then(([doc, references, relationTypes, connectionsGroups]) => {
    // TEST!!!
    const filteredTemplates = connectionsGroups.reduce((templateIds, group) => {
      return templateIds.concat(group.templates.map(t => t._id.toString()));
    }, []);

    const sortOptions = prioritySortingCriteria({currentCriteria: {}, filteredTemplates, templates: globalResources.templates});

    return Promise.all([doc, references, relationTypes, connectionsGroups, referencesAPI.search(documentId, sortOptions), sortOptions]);
    // ---
  })
  .then(([doc, references, relationTypes, connectionsGroups, searchResults, sort]) => {
    return {
      documentViewer: {
        doc,
        references: referencesUtils.filterRelevant(references, lang),
        relationTypes
      },
      // TEST!!!
      connectionsList: {
        entityId: doc.sharedId,
        entityMierda: doc.sharedId,
        entity: doc,
        connectionsGroups,
        searchResults,
        sort,
        filters: {},
        view: 'graph'
      },
      // ----------
      relationTypes
    };
  });
}

export function setViewerState(state) {
  return function (dispatch) {
    const {documentViewer} = state;
    dispatch(actions.set('relationTypes', state.relationTypes));
    dispatch(actions.set('viewer/doc', documentViewer.doc));
    dispatch(actions.set('viewer/relationTypes', documentViewer.relationTypes));
    dispatch(setReferences(documentViewer.references));

    // TEST!!!
    dispatch(actions.set('connectionsList/entityId', state.connectionsList.entityId));
    dispatch(actions.set('connectionsList/entity', state.connectionsList.entity));
    dispatch(actions.set('connectionsList/connectionsGroups', state.connectionsList.connectionsGroups));
    dispatch(actions.set('connectionsList/searchResults', state.connectionsList.searchResults));
    dispatch(actions.set('connectionsList/filters', state.connectionsList.filters));
    dispatch(formActions.merge('connectionsList.sort', state.connectionsList.sort));
    dispatch(actions.set('connectionsList/view', state.connectionsList.view));
  };
}
