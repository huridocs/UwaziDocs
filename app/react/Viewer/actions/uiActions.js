import { actions } from 'app/BasicReducer';
import { events } from 'app/utils';
import { setTargetSelection } from 'app/Viewer/actions/selectionActions';
import Marker from 'app/Viewer/utils/Marker.js';
import scroller from 'app/Viewer/utils/Scroller';
import * as types from 'app/Viewer/actions/actionTypes';

export function closePanel() {
  return {
    type: types.CLOSE_PANEL
  };
}

export function openPanel(panel) {
  return {
    type: types.OPEN_PANEL,
    panel
  };
}

export function resetReferenceCreation() {
  return function (dispatch) {
    dispatch({ type: types.RESET_REFERENCE_CREATION });
    dispatch(actions.unset('viewer/targetDoc'));
    dispatch(actions.unset('viewer/targetDocHTML'));
    dispatch(actions.unset('viewer/targetDocReferences'));
  };
}

export function selectTargetDocument(id) {
  return {
    type: types.SELECT_TARGET_DOCUMENT,
    id
  };
}

export function highlightReference(reference) {
  return {
    type: types.HIGHLIGHT_REFERENCE,
    reference
  };
}

export function deactivateReference() {
  return {
    type: types.DEACTIVATE_REFERENCE
  };
}

export function goToActive(value = true) {
  return {
    type: types.GO_TO_ACTIVE,
    value
  };
}

export function highlightSnippet(snippet) {
  Marker.init('.document-viewer');
  Marker.unmark();
  const page = snippet.get('page');
  Marker.init(`#page-${page}`);
  const text = snippet.get('text');
  if (!text) {
    return;
  }

  const textToMatcherRegExp = _text => _text
  .replace(/[-[\]{}()*+?.,\\^$|#]/g, '\\$&')
  .replace(/<[^>]*>/g, '')
  .replace(/\s+/g, '\\s*')
  .replace(/\n/g, '\\s*');

  const matches = text.match(/<b>(.*?)<\/b>/g).map(m => m.replace(/<.*?>/g, ''));
  const highlight = textToMatcherRegExp(text);

  const scrollToMark = () => {
    scroller.to('.document-viewer mark', '.document-viewer', { duration: 50 });
  };
  const markSearchTerm = () => {
    Marker.init('mark');
    Marker.mark(matches, { className: 'searchTerm', diacritics: false, acrossElements: true, done: scrollToMark });
  };

  const tryFuzziMark = (chunkLenght = 20) => {
    if (!chunkLenght) {
      return;
    }
    const startOfText = textToMatcherRegExp(text.substring(0, chunkLenght));
    const endOfText = textToMatcherRegExp(text.substring(text.length - chunkLenght - 1, text.length - 1));
    const fuzziText = `${startOfText}.*?${endOfText}`;
    const regexp = new RegExp(fuzziText);
    Marker.markRegExp(regexp, {
      separateWordSearch: false,
      acrossElements: true,
      done: markSearchTerm,
      noMatch: tryFuzziMark.bind(this, chunkLenght - 5)
    });
  };

  const regexp = new RegExp(highlight);
  Marker.markRegExp(regexp, {
    separateWordSearch: false,
    acrossElements: true,
    done: markSearchTerm,
    noMatch: tryFuzziMark.bind(this, 20)
  });
}

export function scrollToPage(page, duration = 0) {
  scroller.to(`.document-viewer div#page-${page}`, '.document-viewer', { duration, dividerOffset: 1 });
}

export function scrollTo(reference, docInfo, element = 'a') {
  //
  const page = Object.keys(docInfo).find(pageNumber => docInfo[pageNumber].chars >= reference.range.start);
  //

  if (window.document.querySelector(`.document-viewer ${element}[data-id="${reference._id}"]`, '.document-viewer')) {
    scroller.to(`.document-viewer a[data-id="${reference._id}"]`, '.document-viewer', { duration: 100 });
  } else {
    const scroll = scroller.to(`.document-viewer div#page-${page}`, '.document-viewer', { duration: 0, dividerOffset: 1 });

    events.on('referenceRendered', (renderedReference) => {
      if (renderedReference._id === reference._id &&
          window.document.querySelector(`.document-viewer ${element}[data-id="${reference._id}"]`, '.document-viewer')
      ) {
        window.clearInterval(scroll);
        scroller.to(`.document-viewer ${element}[data-id="${reference._id}"]`, '.document-viewer', { duration: 100 });
        events.removeAllListeners('referenceRendered');
      }
    });
  }

  scroller.to(`.metadata-sidepanel .item-${reference._id}`, '.metadata-sidepanel .sidepanel-body', { duration: 100 });
}

export function selectSnippet(page, snippet) {
  scrollToPage(page);
  return function (dispatch) {
    dispatch({ type: types.SELECT_SNIPPET, snippet });
  };
}

export function activateReference(reference, docInfo, tab) {
  const tabName = tab && !Array.isArray(tab) ? tab : 'references';
  events.removeAllListeners('referenceRendered');

  return function (dispatch) {
    dispatch({ type: types.ACTIVE_REFERENCE, reference: reference._id });
    dispatch({ type: types.OPEN_PANEL, panel: 'viewMetadataPanel' });
    dispatch(actions.set('viewer.sidepanel.tab', tabName));

    setTimeout(() => {
      scrollTo(reference, docInfo);
    });
  };
}

export function scrollToActive(reference, docInfo, tab, doScroll) {
  return function (dispatch) {
    if (doScroll) {
      dispatch(goToActive(false));
      dispatch(activateReference(reference, docInfo, tab));
    }
  };
}


export function selectReference(reference, docInfo) {
  return function (dispatch) {
    dispatch(activateReference(reference, docInfo));
    dispatch(setTargetSelection(reference.range));
  };
}
