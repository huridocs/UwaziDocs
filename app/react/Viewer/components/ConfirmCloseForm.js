import React, {Component, PropTypes} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {actions as formActions} from 'react-redux-form';

import {closePanel} from '../actions/uiActions';
import {hideModal} from 'app/Modals/actions/modalActions';
import Modal from 'app/Layout/Modal';

export class ConfirmCloseForm extends Component {

  confirm() {
    this.props.hideModal('ConfirmCloseForm');
    this.props.closePanel();
    this.props.resetForm('documentViewer.docForm');
  }

  render() {
    if (!this.props.doc) {
      return false;
    }

    let doc = this.props.doc;

    return (
      <Modal isOpen={!!doc} type="danger">

        <Modal.Body>
          <h4>Confirm</h4>
          <p>All changes will be lost, are you sure you want to proceed?</p>
        </Modal.Body>

        <Modal.Footer>
          <button type="button" className="btn btn-default cancel-button" onClick={() => this.props.hideModal('ConfirmCloseForm')}>
            <i className="fa fa-close"></i> Cancel
          </button>
          <button type="button" className="btn btn-danger confirm-button" onClick={() => this.confirm()}>
            <i className="fa fa-trash"></i> Ok
          </button>
        </Modal.Footer>

      </Modal>
    );
  }
}

ConfirmCloseForm.propTypes = {
  hideModal: PropTypes.func,
  resetForm: PropTypes.func,
  closePanel: PropTypes.func,
  doc: PropTypes.object
};

const mapStateToProps = (state) => {
  return {doc: state.modals.get('ConfirmCloseForm')};
};

function mapDispatchToProps(dispatch) {
  return bindActionCreators({closePanel, hideModal, resetForm: formActions.reset}, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(ConfirmCloseForm);
