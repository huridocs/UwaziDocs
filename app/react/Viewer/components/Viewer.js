import PropTypes from 'prop-types';
import React, {Component} from 'react';
import {connect} from 'react-redux';
import Helmet from 'react-helmet';
import {bindActionCreators} from 'redux';

import ContextMenu from 'app/ContextMenu';

import {loadDefaultViewerMenu, loadTargetDocument} from '../actions/documentActions';
import {openPanel} from '../actions/uiActions';
import {addReference} from '../actions/referencesActions';
import SourceDocument from './SourceDocument';
import TargetDocument from './TargetDocument';
import {CreateConnectionPanel} from 'app/Connections';
import ViewMetadataPanel from './ViewMetadataPanel';

import ViewerDefaultMenu from './ViewerDefaultMenu';
import ViewerTextSelectedMenu from './ViewerTextSelectedMenu';
import ConfirmCloseForm from './ConfirmCloseForm';
import Footer from 'app/App/Footer';
import ShowIf from 'app/App/ShowIf';
import {TemplateLabel, Icon} from 'app/Layout';
import Marker from 'app/Viewer/utils/Marker';

import {ConnectionsList} from 'app/ConnectionsList';
import RelationshipMetadata from 'app/Relationships/components/RelationshipMetadata';

export class Viewer extends Component {

  componentWillMount() {
    this.context.store.dispatch(openPanel('viewMetadataPanel'));
  }

  componentDidMount() {
    this.context.store.dispatch(loadDefaultViewerMenu());
    Marker.init('div.main-wrapper');
  }

  render() {
    const {doc, sidepanelTab} = this.props;

    let className = 'document-viewer';
    if (this.props.panelIsOpen) {
      className += ' with-panel is-active';
    }
    if (this.props.targetDoc) {
      className += ' show-target-document';
    }
    if (this.props.showConnections) {
      className += ' connections';
    }

    return (
      <div className="row">
        <Helmet title={doc.get('title') ? doc.get('title') : 'Document'} />
        <ShowIf if={!this.props.targetDoc}>
          <div className="content-header content-header-document">
            <div className="content-header-title">
              <Icon className="item-icon item-icon-center" data={doc.get('icon') ? doc.get('icon').toJS() : {}} size="sm"/>
              <h1 className="item-name">{doc.get('title')}</h1>
              <TemplateLabel template={doc.get('template')}/>
            </div>
          </div>
        </ShowIf>
        <main className={className}>
          <div className="main-wrapper">
            <ShowIf if={sidepanelTab !== 'connections' && !this.props.targetDoc}>
              <SourceDocument page={this.props.page} searchTerm={this.props.searchTerm}/>
            </ShowIf>
            <ShowIf if={sidepanelTab === 'connections'}>
              <ConnectionsList deleteConnection={() => {}} />
            </ShowIf>
            <TargetDocument />
            <Footer/>
          </div>
        </main>

        <ConfirmCloseForm />
        <ViewMetadataPanel storeKey={'documentViewer'} searchTerm={this.props.searchTerm}/>
        <CreateConnectionPanel containerId={this.props.targetDoc ? 'target' : doc.get('sharedId')}
                               onCreate={this.props.addReference}
                               onRangedConnect={this.props.loadTargetDocument} />
        <ShowIf if={sidepanelTab === 'connections'}>
          <RelationshipMetadata />
        </ShowIf>

        <ContextMenu align="bottom" overrideShow={true} show={!this.props.panelIsOpen}>
          <ViewerDefaultMenu/>
        </ContextMenu>
        <ContextMenu align="center" overrideShow={true} show={this.props.showTextSelectMenu}>
          <ViewerTextSelectedMenu/>
        </ContextMenu>
      </div>
    );
  }
}

Viewer.propTypes = {
  doc: PropTypes.object,
  searchTerm: PropTypes.string,
  page: PropTypes.number,
  panelIsOpen: PropTypes.bool,
  addReference: PropTypes.func,
  targetDoc: PropTypes.bool,
  // TEST!!!
  sidepanelTab: PropTypes.string,
  loadTargetDocument: PropTypes.func,
  showConnections: PropTypes.bool,
  showTextSelectMenu: PropTypes.bool,
  selectedConnection: PropTypes.bool,
  selectedConnectionMetadata: PropTypes.object
};

Viewer.contextTypes = {
  store: PropTypes.object
};

const mapStateToProps = (state) => {
  let documentViewer = state.documentViewer;
  let uiState = documentViewer.uiState.toJS();
  return {
    doc: documentViewer.doc,
    panelIsOpen: !!uiState.panel,
    targetDoc: !!documentViewer.targetDoc.get('_id'),
    // TEST!!!
    sidepanelTab: documentViewer.sidepanel.tab,
    showConnections: documentViewer.sidepanel.tab === 'references',
    showTextSelectMenu: Boolean(!documentViewer.targetDoc.get('_id') && uiState.reference && uiState.reference.sourceRange)
  };
};

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators({addReference, loadTargetDocument}, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(Viewer);
