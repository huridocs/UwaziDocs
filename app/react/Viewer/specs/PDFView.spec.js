/**
 * @jest-environment jsdom
 */
/* eslint-disable max-statements */
/* eslint-disable react/no-multi-comp */
import React from 'react';
import { fromJS } from 'immutable';
import entitiesAPI from 'app/Entities/EntitiesAPI';
import { actions } from 'app/BasicReducer';
import { PDFView } from 'app/Viewer/PDFView';
import { ConnectedViewer as Viewer } from 'app/Viewer/components/Viewer';
import RouteHandler from 'app/App/RouteHandler';
import * as utils from 'app/utils';
import { RequestParams } from 'app/utils/RequestParams';
import { renderConnectedMount } from 'app/utils/test/renderConnected';
import * as documentActions from 'app/Viewer/actions/documentActions';
import * as routeActions from '../actions/routeActions';
import * as uiActions from '../actions/uiActions';

let page = 0;
let raw = '';
let pathname = '';
let ref = '';
let anotherProp = '';

const mockUseLocation = jest.fn().mockImplementation(() => ({
  search: `?raw=${raw},page=${page}`,
  pathname,
}));

const mockUseSearchParams = jest.fn().mockImplementation(() => {
  const params = new Map();
  params.set('raw', raw);
  params.set('page', page);
  if (ref) {
    params.set('ref', ref);
  }
  if (anotherProp) {
    params.set('anotherProp', anotherProp);
  }

  return [params];
});
const mockNavigate = jest.fn().mockImplementation(path => path);
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useSearchParams: () => mockUseSearchParams(),
  useLocation: () => mockUseLocation(),
  useNavigate: () => path => mockNavigate(path),
  useMatches: () => {},
}));
jest.mock('app/ContextMenu', () => () => <div>ContextMenu</div>);
jest.mock('app/App/Footer', () => () => <div>Footer</div>);
jest.mock('app/Viewer/components/ViewMetadataPanel', () => () => <div>ViewMetadataPanel</div>);
jest.mock('app/Connections', () => ({
  CreateConnectionPanel: () => <div>CreateConnectionPanel</div>,
}));

describe('PDFView', () => {
  let component;
  let instance;
  let context;
  let props;

  const state = {
    documentViewer: {
      uiState: fromJS({ reference: { targetRange: [] } }),
      targetDoc: fromJS({ _id: 'document1' }),
      sidepanel: { tab: '1', metadata: { _id: 'prop1' } },
      targetDocReferences: [],
    },
    connections: { connection: fromJS({}) },
    user: fromJS({ _id: 'user1' }),
    settings: { collection: fromJS({}) },
    modals: fromJS({ ConfirmCloseForm: () => {} }),
    semanticSearch: { selectedDocument: fromJS({}) },
  };

  const render = () => {
    RouteHandler.renderedFromServer = true;
    component = renderConnectedMount(PDFView, state, props, true, { context });
    //component.setContext({ ...component.context(), ...context });
    instance = component.instance();
    component.instance().getChildContext().store.dispatch = context.store.dispatch;
  };

  beforeEach(() => {
    page = 1;
    raw = '';
    pathname = '';
    ref = '';
    anotherProp = '';

    const dispatch = jasmine.createSpy('dispatch');
    context = {
      store: {
        getState: () => ({}),
        dispatch: dispatch.and.callFake(action =>
          typeof action === 'function' ? action(dispatch) : action
        ),
        subscribe: jasmine.createSpy('subscribe'),
      },
    };

    props = {
      entity: fromJS({ defaultDoc: { _id: 'documentId' } }),
      location: { query: {} },
      routes: [],
    };

    render();

    spyOn(routeActions, 'requestViewerState');

    spyOn(routeActions, 'setViewerState').and.returnValue({ type: 'setViewerState' });
  });

  it('should pass down raw property', () => {
    raw = 'true';
    page = 2;
    render();
    expect(component.find(Viewer).props().raw).toEqual(true);
  });

  describe('when on server', () => {
    it('should always pass raw true', () => {
      raw = 'false';
      // eslint-disable-next-line no-import-assign
      utils.isClient = false;
      render();
      expect(component.find(Viewer).props().raw).toBe(true);
      // eslint-disable-next-line no-import-assign
      utils.isClient = true;
    });
  });

  it('should render the Viewer', () => {
    expect(component.find(Viewer).length).toBe(1);
  });

  describe('when raw', () => {
    it('should render link canonical to the not raw version', () => {
      raw = 'true';
      page = 1;
      pathname = 'pathname';
      render();
      expect(
        component.find({ link: [{ href: 'pathname?page=1', rel: 'canonical' }] })
      ).toMatchSnapshot();
    });
  });
  describe('when not raw', () => {
    it('should not render link canonical', () => {
      raw = 'false';
      page = 1;
      pathname = 'pathname';
      render();
      expect(component.find({ link: [{ href: 'pathname?page=1', rel: 'canonical' }] }).length).toBe(
        0
      );
    });
  });
  describe('static requestState', () => {
    it('should call on requestViewerState', () => {
      const requestParams = new RequestParams({
        documentId: 'documentId',
        lang: 'es',
        page: 4,
        raw: 'true',
      });
      PDFView.requestState(requestParams, 'globalResources');
      expect(routeActions.requestViewerState).toHaveBeenCalledWith(
        new RequestParams({ documentId: 'documentId', lang: 'es', raw: true, page: 4 }),
        'globalResources'
      );
    });

    it('should modify raw to true if is server side rendered', () => {
      // eslint-disable-next-line no-import-assign
      utils.isClient = false;
      const requestParams = new RequestParams({
        documentId: 'documentId',
        lang: 'es',
        raw: 'false',
      });
      PDFView.requestState(requestParams, 'globalResources');
      expect(routeActions.requestViewerState).toHaveBeenCalledWith(
        new RequestParams({ documentId: 'documentId', lang: 'es', raw: true }),
        'globalResources'
      );
    });
  });

  describe('onDocumentReady', () => {
    it('should scrollToPage on the query when not on raw mode', () => {
      spyOn(uiActions, 'scrollToPage');
      raw = 'false';
      page = 15;
      render();
      component.find({ page: 15 }).at(0).props().onDocumentReady();
      expect(uiActions.scrollToPage).toHaveBeenCalledWith(15, 0);

      raw = 'true';
      render();
      uiActions.scrollToPage.calls.reset();
      component.find({ page: 15 }).at(0).props().onDocumentReady();
      expect(uiActions.scrollToPage).not.toHaveBeenCalled();
    });

    it('should activate text reference if query parameters have reference id', () => {
      spyOn(uiActions, 'activateReference');
      raw = 'false';
      ref = 'refId';
      pathname = 'pathname';
      const reference = { _id: 'refId', range: { start: 200, end: 300 }, text: 'test' };
      const doc = fromJS({
        relations: [{ _id: 'otherRef' }, reference],
      });
      render();
      component.find({ page: 1 }).at(0).props().onDocumentReady(doc);
      expect(uiActions.activateReference).toHaveBeenCalledWith(reference);
    });

    it('should emit documentLoaded event', () => {
      spyOn(uiActions, 'scrollToPage');
      spyOn(utils.events, 'emit');
      render();

      component.find({ page: 1 }).at(0).props().onDocumentReady();
      expect(utils.events.emit).toHaveBeenCalledWith('documentLoaded');
    });
  });

  describe('changePage', () => {
    describe('when raw', () => {
      it('should changeBrowserHistoryPage', () => {
        raw = 'true';
        page = 15;
        anotherProp = 'test';
        pathname = 'pathname';
        spyOn(uiActions, 'scrollToPage');
        render();

        component.find({ page: 15 }).at(0).props().changePage(16);
        expect(mockNavigate).toHaveBeenCalledWith('pathname?raw=true&anotherProp=test&page=16');
        expect(uiActions.scrollToPage).not.toHaveBeenCalled();
      });
    });

    describe('when not raw', () => {
      it('should scrollToPage', () => {
        raw = 'false';
        page = 15;
        anotherProp = 'test';
        pathname = 'pathname';
        spyOn(uiActions, 'scrollToPage');
        mockNavigate.mockClear();

        render();
        component.find({ page: 15 }).at(0).props().changePage(16);
        expect(mockNavigate).not.toHaveBeenCalled();
        expect(uiActions.scrollToPage).toHaveBeenCalledWith(16);
      });
    });
  });

  describe('componentWillReceiveProps', () => {
    const setProps = newProps => {
      entitiesAPI.getRawPage.calls.reset();
      component.setProps(newProps);
      component.update();
    };

    beforeEach(() => {
      props.params = { sharedId: 'entityId' };
      spyOn(entitiesAPI, 'getRawPage').and.returnValue(Promise.resolve('raw text'));
      render();
    });

    it('should load raw page when page/raw changes and raw is true', async () => {
      raw = 'true';
      page = 15;
      expect(entitiesAPI.getRawPage).not.toHaveBeenCalled();

      setProps({ location: { query: { page: 16, raw: 'false' } } });
      expect(entitiesAPI.getRawPage).not.toHaveBeenCalled();

      entitiesAPI.getRawPage.calls.reset();
      await instance.componentWillReceiveProps({
        params: { sharedId: 'entityId' },
        location: { query: { page: 17, raw: 'true' } },
        entity: fromJS({ defaultDoc: { _id: 'documentId' } }),
      });
      expect(context.store.dispatch).toHaveBeenCalledWith(
        actions.set('viewer/rawText', 'raw text')
      );
      expect(entitiesAPI.getRawPage).toHaveBeenCalledWith(
        new RequestParams({ _id: 'documentId', page: 17 })
      );
    });
  });

  describe('changeBrowserHistoryPage', () => {
    it('should push new browserHistory with new page', () => {
      raw = 'true';
      page = 15;
      anotherProp = 'test';
      pathname = 'pathname';
      render();

      component.find({ page: 15 }).at(0).props().changePage(16);

      expect(mockNavigate).toHaveBeenCalledWith('pathname?raw=true&anotherProp=test&page=16');

      raw = 'false';
      page = 15;
      anotherProp = 'test';
      mockNavigate.mockClear();
      component.find(PDFView).update();

      component.find({ page: 15 }).at(0).props().changePage(16);

      expect(mockNavigate).toHaveBeenCalledWith('pathname?page=16');
    });
  });

  describe('componentWillUnmount', () => {
    it('should leave edit mode', () => {
      spyOn(documentActions, 'leaveEditMode').and.returnValue({ type: 'LEAVING_EDIT_MODE' });
      render();
      component.unmount();
      expect(documentActions.leaveEditMode).toHaveBeenCalled();
    });
  });
});
