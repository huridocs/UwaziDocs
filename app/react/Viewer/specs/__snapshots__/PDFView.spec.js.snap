// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PDFView when raw should render link canonical to the not raw version 1`] = `
Array [
  <SideEffect(NullComponent)
    defer={true}
    encodeSpecialCharacters={true}
    link={
      Array [
        Object {
          "href": "pathname?page=1",
          "rel": "canonical",
        },
      ]
    }
  >
    <NullComponent
      defer={true}
      encodeSpecialCharacters={true}
      link={
        Array [
          Object {
            "href": "pathname?page=1",
            "rel": "canonical",
          },
        ]
      }
    />
  </SideEffect(NullComponent)>,
  <NullComponent
    defer={true}
    encodeSpecialCharacters={true}
    link={
      Array [
        Object {
          "href": "pathname?page=1",
          "rel": "canonical",
        },
      ]
    }
  />,
]
`;

exports[`PDFView when raw when page is undefined should render link canonical with a default page 1`] = `
Array [
  <SideEffect(NullComponent)
    defer={true}
    encodeSpecialCharacters={true}
    link={
      Array [
        Object {
          "href": "pathname?page=1",
          "rel": "canonical",
        },
      ]
    }
  >
    <NullComponent
      defer={true}
      encodeSpecialCharacters={true}
      link={
        Array [
          Object {
            "href": "pathname?page=1",
            "rel": "canonical",
          },
        ]
      }
    />
  </SideEffect(NullComponent)>,
  <NullComponent
    defer={true}
    encodeSpecialCharacters={true}
    link={
      Array [
        Object {
          "href": "pathname?page=1",
          "rel": "canonical",
        },
      ]
    }
  />,
]
`;
